controller/konnect/ops/ops_controlplane.go

```diff
@@ -34,6 +35,36 @@ func convertCreateControlPlaneRequestToSDK(
 	}
 }
 
+// ensureControlPlane ensures that the Konnect ControlPlane exists in Konnect. It is created
+// if it does not exist and the source is Origin. If the source is Mirror, it checks
+// if the ControlPlane exists in Konnect and returns an error if it does not.
+func ensureControlPlane(
+	ctx context.Context,
+	sdk sdkops.ControlPlaneSDK,
+	sdkGroups sdkops.ControlPlaneGroupSDK,
+	cl client.Client,
+	cp *konnectv1alpha1.KonnectGatewayControlPlane,
+) error {
+	switch *cp.Spec.Source {
+	case commonv1alpha1.EntitySourceOrigin:
+		return createControlPlane(ctx, sdk, sdkGroups, cl, cp)
+	case commonv1alpha1.EntitySourceMirror:
+		if _, err := GetControlPlaneByID(
+			ctx,
+			sdk,
+			// not nilness is ensured by CEL rules
+			string(cp.Spec.Mirror.Konnect.ID),
+		); err != nil {
+			return err
+		}
+		cp.SetKonnectID(string(cp.Spec.Mirror.Konnect.ID))
+		return nil
+	default:
+		// This should never happen, as the source type is validated by CEL rules.
```

Comment:
There would be legacy cps that does not have this field so this may happen. Could you please test running it with cps created based on the older version of CRD without the `spec.source` field? This may bring breaking changes,

Reply:
It cannot happen, as the source field is defaulted to `Source` if it does not exist. When upgrading the CRDs, the old CPs get automatically updated to be configured as source. For this reason, this is not a breaking change.